modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/SnapshotCheckProcess.java

```diff
@@ -0,0 +1,549 @@
+    }
+
+    /** */
+    private Map<ClusterNode, Map<PartitionKeyV2, PartitionHashRecordV2>> collectPartsHashes(
+        @Nullable Map<UUID, CheckResultDTO> results,
+        Collection<UUID> requiredNodes
+    ) {
+        if (F.isEmpty(results))
+            return Collections.emptyMap();
+
+        return results.entrySet().stream()
+            .filter(e -> requiredNodes.contains(e.getKey()) && e.getValue() != null)
+            .collect(Collectors.toMap(e -> kctx.cluster().get().node(e.getKey()), e -> e.getValue().partsHashes));
+    }
+
+    /**
+     * @param snpName Snapshot name of the validation process. If {@code null}, ignored.
+     * @param procId  If {@code snpName} is {@code null}, is used to find the operation request.
+     * @return Current snapshot checking request by {@code snpName} or {@code procId}.
+     */
+    private @Nullable SnapshotCheckProcessRequest currentRequest(@Nullable String snpName, UUID procId) {
+        return snpName == null
+            ? requests.values().stream().filter(rq -> rq.requestId().equals(procId)).findFirst().orElse(null)
+            : requests.get(snpName);
+    }
+
+    /** Phase 1 beginning: prepare, collect and check local metas. */
+    private IgniteInternalFuture<CheckResultDTO> prepareAndCheckMetas(SnapshotCheckProcessRequest extReq) {
+        SnapshotCheckProcessRequest locReq = requests.computeIfAbsent(extReq.snapshotName(), snpName -> extReq);
+
+        if (!locReq.equals(extReq)) {
+            Throwable err = new IllegalStateException("Validation of snapshot '" + extReq.snapshotName()
+                + "' has already started. Request=" + locReq + '.');
+
+            clean(extReq.requestId(), err, null, null);
+
+            return new GridFinishedFuture<>(err);
+        }
+
+        if (!extReq.nodes.contains(kctx.localNodeId())) {
+            if (log.isDebugEnabled()) {
+                log.debug("Skipping snapshot local metadatas collecting for snapshot validation, request=" + extReq
+                    + ". Current node is not required.");
+            }
+
+            return new GridFinishedFuture<>();
+        }
+
+        GridFutureAdapter<CheckResultDTO> locMetasChkFut = new GridFutureAdapter<>();
+
+        stopFutureOnAnyFailure(locMetasChkFut, () -> {
+            assert locReq.fut() == null;
+
+            locReq.fut(locMetasChkFut);
+
+            IgniteSnapshotManager snpMgr = kctx.cache().context().snapshotMgr();
+
+            if (log.isDebugEnabled())
+                log.debug("Checking local snapshot metadatas. Request=" + locReq + '.');
+
+            Collection<Integer> grpIds = F.isEmpty(locReq.groups()) ? null : F.viewReadOnly(locReq.groups(), CU::cacheId);
+
+            // An error can occur when the local future is not initialized yet.
+            Throwable locReqErr = locReq.error();
+
+            if (locReqErr != null)
+                locMetasChkFut.onDone(locReqErr);
+
+            if (!locMetasChkFut.isDone()) {
```

Comment:
Actually, do we really need this check at all? The future inside `checkLocalMetas` isn't interruptible, then in case of error happened a moment later after the check, it unconditionally checks all metas. I think we can skip this `if` at all in favor of simplicity.

Reply:
I think we should not start any working if we cleary know that an error already arised. Why not checking it before starting?

Reply:
> Why not checking it before starting?

Because you don't check it after start, for example for every iteration of underlying loop inside `checkLocalMetas`. You should check it everywhere to reduce amount of work to be done. Currently it looks weird and only complicates the code.

This exception is throwed by very rare condition and amount of job to be skipped is actually very small. Let's skip this handling of this case.

Reply:
Still not fixed